libhb/avfilter.c

```diff
@@ -237,17 +237,25 @@ void hb_avfilter_alias_close( hb_filter_object_t * filter )
 static hb_buffer_t* filterFrame( hb_filter_private_t * pv, hb_buffer_t * in )
 {
     hb_buffer_list_t   list;
-    hb_buffer_t      * buf, * next;
+    hb_buffer_t      * buf = NULL, * next = NULL;
+    int av_frame_is_not_null = 1; // TODO: find the reason for emply input av_frame for ffmpeg filters
```

Comment:
When the final "end of file" mark is reached, a NULL frame is given to the avfilter graph to flush any delayed frames. hb_avfilter_add_buf calls av_buffersrc_add_frame with a NULL buffer which should be allowed in this case by avfilter.

I need more time to review this PR.  It's not clear to me why you would want to call hb_avfilter_add_buf at all when QSV filters are enabled.  Is avfilter somehow involved in QSV filtering?  Can you give me a brief tutorial on the flow packets follow when qsv filtering is enabled please?

I'm also not at all thrilled about the addition of the global variable qsv_filters_are_enabled.  Would it be possible to add this as a field of hb_job_t instead of using a global?  job appears to be passed to most places this is needed.

Reply:
Actually, now that I think of it, use of a global *will* break when 2 encodes are happening simultaneously (i.e. live preview and regular encode).  So this needs to be fixed.  If there are any additional globals in the qsv encode code, these also need to be fixed.  I've never looked all that closely at the qsv code so could have missed other global usage.

Just an FYI, there is currently also this kind of global variable abuse in x265 encodes because *they* also can't seem to keep all parameters isolated in the encode context.  It's one of the reasons the windows gui has been adding process isolation (maybe mac UI has as well?).  But the Linux GUI hasn't done this yet.

Reply:
This is not the case that you described. 

This happens when we start processing at the first frame by av_filter, not in the end. Moreover the in hb_buffer is not null always, but in->qsv_details.frame is null. It means that buffer came not from the QSV decoder! Moreover it happens only when GUI binary is used, CLI works fine. It is definitely unexpected behavior.

hb_avfilter_add_buf is needed because we do scaling via FFMPEG. VPP filters are implemented in FFMPEG.
https://github.com/FFmpeg/FFmpeg/blob/master/libavfilter/vf_scale_qsv.c

Agree about global variables. It will be refactored soon.

Reply:
good point about globals, @jstebbins 

FFMPEG based VPP/resize works 1-to-1 , no flashing should be considered,
as @agalin89 wrote before - this implementation re-uses QSV implementation from FFMPEG rather than own.
it should be long term and easy in support solution.